Opts in lib

alanz opened the pull request:
Note: hlint has changed `Error` to `Warning` in its output, some tests fail for that.

alanz committed:
Move the CLI parser definitions into the lib

So they can be used by library clients to parse and pass in the
appropriate ghc-mod Options.

alanz committed:
Fix tests, export parser spec

lierdakil commented:
I'm not too keen on this version. E.g.  `parseArgs` and `parseArgsInteractive` make literally no sense in library context. A proper library interface might be in order. You might want to look to hlint for inspiration.

If you only want to parse `Options`, you only need to get `globalArgsSpec` to library, not _everything_.

alanz commented:
Yes, `globalArgsSpec` is all I need, but does it makes sense to have that in the library and the rest not? I did similar for HaRe initially, and there is quite a bit of that stuff that has to come along. See https://github.com/alanz/HaRe/blob/wip/src/Options.hs

lierdakil commented:
`Commands.hs` makes no sense in there, since those definitions aren't used outside of executable. `ShellParse.hs` doesn't make sense there, since it's used only by executable. Why would you even need those in library is beyond me, frankly.

alanz commented:
Ok, I will update the PR, no problem.

On Tue, Feb 9, 2016 at 3:43 PM, Nikolay Yakimov notifications@github.com
wrote:

> Commands.hs makes no sense in there, since those definitions aren't used
> outside of executable. ShellParse.hs doesn't make sense there, since it's
> used only by executable. Why would you even need those in library is beyond
> me, frankly.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/DanielG/ghc-mod/pull/745#issuecomment-181868968.

alanz committed:
Merge branch 'master' into opts-in-lib

alanz committed:
Rebalance between lib and exe

alanz committed:
Cleaning up

alanz commented:
@lierdakil How does that look?

lierdakil commented:
Sorry, not much time atm, I'll make sure to take a look once I get some
time. Probably by weekend.

2016-02-09 23:55 GMT+03:00 Alan Zimmerman notifications@github.com:

> @lierdakil https://github.com/lierdakil How does that look?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/DanielG/ghc-mod/pull/745#issuecomment-182063363.

alanz commented:
ok, not urgent

On Wed, Feb 10, 2016 at 10:47 PM, Nikolay Yakimov notifications@github.com
wrote:

> Sorry, not much time atm, I'll make sure to take a look once I get some
> time. Probably by weekend.
> 
> 2016-02-09 23:55 GMT+03:00 Alan Zimmerman notifications@github.com:
> 
> > @lierdakil https://github.com/lierdakil How does that look?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/DanielG/ghc-mod/pull/745#issuecomment-182063363.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/DanielG/ghc-mod/pull/745#issuecomment-182577086.

lierdakil commented:
This is better. I think you also might want to add an utility function to `Language.Haskell.GhcMod.Options.Options`, along the lines of `parseCmdLine :: [String] -> Options`, so that library users don't _have_ to depend on optparse-applicative directly. It also might be a good idea to expose that function in `Language.Haskell.GhcMod`.

Also, do add haddock documentation for library exports. It bears at least some explanation, I think.

alanz committed:
Merge remote-tracking branch 'dxld/master' into opts-in-lib

alanz committed:
Add a argument parser and some haddocks

I am not sure if I have set the ParserInfo up correctly, I have never
used it before.

alanz commented:
@DanielG @lierdakil Any reason not to merge this?

lierdakil commented:
Sorry, I was kinda out of it for a bit. `execParserMaybe` is deprecated. Use `execParserPure` and pass its result to `getParseResult` to emulate the same behavior (or you might pattern-match on `ParseResult` and return an `Either`, but that would require some black magic with `renderFailure` to be useful)

Also, you don't need to fill in `ParserInfo` by hand. Just use `info` from `Options.Applicative.Builder` and pass it a `Parser` and `mempty` to generate default `ParserInfo`.

An example from the top of my head:

``` haskell
-- all options are set to defaults
parseCmdLineOptions :: [String] -> Maybe Options
parseCmdLineOptions = getParseResult . execParserPure (prefs mempty) (info globalArgSpec mempty)
```

lierdakil commented:
Also bear in mind that pushes are not sent via e-mail, so I can miss updates if there are no follow-up comments. Dropping a line in comment after pushing might save us both some time :wink:

alanz commented:
Thanks for the comments. I have never used the execParserXXX stuff so I was winging it, and did see that it was deprecated. I was hoping for some constructive feedback which you have provided.

alanz committed:
Merge branch 'master' into opts-in-lib

alanz committed:
Do not use deprecated API for parseCmdLineOptions

Thanks to @lierdakil

alanz commented:
Ping @lierdakil, updated

alanz committed:
Sort out imports for earlier GHC versions

lierdakil commented:
Um. We try to contain CPP infestation by having most CPP stuff in `Language.Haskell.GhcMod.Gap`. Thankfully, you don't really need to use it here.

``` haskell
import Data.Monoid
import Prelude
```

will work with no warnings on all supported GHC versions.

lierdakil commented:
Other than that, LGTM. @DanielG, any reservations against merging this?

DanielG commented:
Nope feel free to merge.

alanz committed:
Remove CPP from Options

As per @lierdakil suggestion

alanz commented:
@DanielG @lierdakil I have made the final update, the tests end up with the same hlint Error vs Warning problem as before, so I think it is ready for a merge.

Thanks for the patience and feedback
